Source/SwiftLintFramework/Rules/GenericTypeNameRule.swift

```diff
@@ -0,0 +1,235 @@
+        guard let beforeWhere = constraint.components(separatedBy: "where").first else {
+            return []
+        }
+
+        let namesAndRanges: [(String, NSRange)] = beforeWhere.split(separator: ",").flatMap { string, range in
+            return string.split(separator: ":").first.map {
+                let (trimmed, trimmedRange) = $0.0.trimmingWhitespaces()
+                return (trimmed, NSRange(location: range.location + trimmedRange.location,
+                                         length: trimmedRange.length))
+            }
+        }
+
+        let contents = file.contents.bridge()
+        return namesAndRanges.flatMap { (name, range) -> (String, Int)? in
+            guard let byteRange = contents.NSRangeToByteRange(start: range.location + offset,
+                                                              length: range.length),
+                case let kinds = file.syntaxMap.tokensIn(byteRange).flatMap({ SyntaxKind(rawValue: $0.type) }),
+                kinds == [.identifier] else {
+                    return nil
+            }
+
+            return (name, byteRange.location)
+        }
+    }
+
+    private func validateName(name: String, file: File, offset: Int) -> [StyleViolation] {
+        guard !configuration.excluded.contains(name) else {
+            return []
+        }
+
+        if !CharacterSet.alphanumerics.isSuperset(ofCharactersIn: name) {
+            return [
+                StyleViolation(ruleDescription: type(of: self).description,
+                               severity: .error,
+                               location: Location(file: file, byteOffset: offset),
+                               reason: "Generic type name should only contain alphanumeric characters: '\(name)'")
+            ]
+        } else if !name.substring(to: name.index(after: name.startIndex)).isUppercase() {
+            return [
+                StyleViolation(ruleDescription: type(of: self).description,
+                               severity: .error,
+                               location: Location(file: file, byteOffset: offset),
+                               reason: "Generic type name should start with an uppercase character: '\(name)'")
+            ]
+        } else if let severity = severity(forLength: name.characters.count) {
+            return [
+                StyleViolation(ruleDescription: type(of: self).description,
+                               severity: severity,
+                               location: Location(file: file, byteOffset: offset),
+                               reason: "Generic type name should be between \(configuration.minLengthThreshold) and " +
+                    "\(configuration.maxLengthThreshold) characters long: '\(name)'")
+            ]
+        }
+
+        return []
+    }
+}
+
+extension String {
+    fileprivate func split(separator: Character) -> [(String, NSRange)] {
```

Comment:
I'd probably still have a helper function to calculate the ranges and return a `[(String, NSRange)]`. I could reimplement this function to use `String.components(separatedBy:)` instead. Do you think it would be better?